File: internal/model/raybot_command.go
package model

import (
	"context"
	"fmt"
	"time"

	"github.com/google/uuid"

	"github.com/tuanvumaihuynh/roboflow/pkg/paging"
	"github.com/tuanvumaihuynh/roboflow/pkg/xerrors"
	"github.com/tuanvumaihuynh/roboflow/pkg/xsort"
)

// RaybotCommandStatus represents the state of a command
type RaybotCommandStatus string

func (s RaybotCommandStatus) Validate() error {
	switch s {
	case RaybotCommandStatusPending,
		RaybotCommandStatusInProgress,
		RaybotCommandStatusSuccess,
		RaybotCommandStatusFailed:
	default:
		return xerrors.ThrowInvalidArgument(nil, fmt.Sprintf("invalid command status: %s", s))
	}
	return nil
}

const (
	RaybotCommandStatusPending    RaybotCommandStatus = "PENDING"
	RaybotCommandStatusInProgress RaybotCommandStatus = "IN_PROGRESS"
	RaybotCommandStatusSuccess    RaybotCommandStatus = "SUCCESS"
	RaybotCommandStatusFailed     RaybotCommandStatus = "FAILED"
)

// RaybotCommandType represents the type of a command
type RaybotCommandType string

func (c RaybotCommandType) Validate() error {
	switch c {
	case RaybotCommandTypeStop,
		RaybotCommandTypeMoveForward,
		RaybotCommandTypeMoveBackward,
		RaybotCommandTypeMoveToLocation,
		RaybotCommandTypeOpenBox,
		RaybotCommandTypeCloseBox,
		RaybotCommandTypeLiftBox,
		RaybotCommandTypeDropBox,
		RaybotCommandTypeCheckQrCode,
		RaybotCommandTypeWaitGetItem,
		RaybotCommandTypeScanLocation:
	default:
		return xerrors.ThrowInvalidArgument(nil, fmt.Sprintf("invalid command type: %s", c))
	}
	return nil
}

const (
	RaybotCommandTypeStop           RaybotCommandType = "STOP"
	RaybotCommandTypeMoveForward    RaybotCommandType = "MOVE_FORWARD"
	RaybotCommandTypeMoveBackward   RaybotCommandType = "MOVE_BACKWARD"
	RaybotCommandTypeMoveToLocation RaybotCommandType = "MOVE_TO_LOCATION"
	RaybotCommandTypeOpenBox        RaybotCommandType = "OPEN_BOX"
	RaybotCommandTypeCloseBox       RaybotCommandType = "CLOSE_BOX"
	RaybotCommandTypeLiftBox        RaybotCommandType = "LIFT_BOX"
	RaybotCommandTypeDropBox        RaybotCommandType = "DROP_BOX"
	RaybotCommandTypeCheckQrCode    RaybotCommandType = "CHECK_QR"
	RaybotCommandTypeWaitGetItem    RaybotCommandType = "WAIT_GET_ITEM"

	// Command that has response data
	RaybotCommandTypeScanLocation RaybotCommandType = "SCAN_LOCATION"
)

type MoveToLocationInput struct {
	Location  string `json:"location" validate:"required"`
	Direction string `json:"direction" validate:"required,oneof=FORWARD BACKWARD"`
}

type CheckQRCodeInput struct {
	QRCode string `json:"qr_code" validate:"required,qrcode"`
}

type LiftDropBoxInput struct {
	Distance *int32 `json:"distance" validate:"omitempty,gte=300,lte=2000"`
}

type FailedOutput struct {
	Reason string `json:"reason"`
}

type ScanLocationOutput struct {
	Locations []string `json:"locations"`
}

type RaybotCommand struct {
	RaybotID    uuid.UUID
	ID          uuid.UUID
	Type        RaybotCommandType
	Status      RaybotCommandStatus
	Input       []byte
	Output      []byte
	CreatedAt   time.Time
	CompletedAt *time.Time
}

type RaybotCommandRepository interface {
	Get(ctx context.Context, id uuid.UUID) (RaybotCommand, error)
	List(ctx context.Context, raybotId uuid.UUID, p paging.Params, sorts []xsort.Sort) (*paging.List[RaybotCommand], error)
	Create(ctx context.Context, cmd RaybotCommand) error

	// Update updates both the raybot command and its associated raybot's status
	Update(
		ctx context.Context,
		cmdID uuid.UUID,
		raybotStatus RaybotStatus,
		fn func(raybotCmd *RaybotCommand) error,
	) error
	Delete(ctx context.Context, id uuid.UUID) error
}

File: ui/src/types/raybot-command.ts
export const RaybotCommandTypeValues = [
  'STOP',
  'MOVE_FORWARD',
  'MOVE_BACKWARD',
  'MOVE_TO_LOCATION',
  'OPEN_BOX',
  'CLOSE_BOX',
  'LIFT_BOX',
  'DROP_BOX',
  'CHECK_QR',
  'WAIT_GET_ITEM',
  'SCAN_QR_LOCATION',
] as const
export type RaybotCommandType = typeof RaybotCommandTypeValues[number]

export const RaybotCommandStatusValues = [
  'PENDING',
  'IN_PROGRESS',
  'SUCCESS',
  'FAILED',
] as const
export type RaybotCommandStatus = typeof RaybotCommandStatusValues[number]

export interface InputMap {
  STOP: undefined
  MOVE_FORWARD: undefined
  MOVE_BACKWARD: undefined
  OPEN_BOX: undefined
  CLOSE_BOX: undefined
  WAIT_GET_ITEM: undefined
  DROP_BOX?: { distance?: number }
  LIFT_BOX?: { distance?: number }
  MOVE_TO_LOCATION: {
    location: string
    direction: 'FORWARD' | 'BACKWARD'
  }
  CHECK_QR: { qr_code: string }
  SCAN_QR_LOCATION: undefined
}

export type SuccessOutputMap = Record<RaybotCommandType, undefined> & {
  SCAN_QR_LOCATION: { locations: string[] }
}

export interface OutputMap {
  FAILED: { reason: string }
  SUCCESS: SuccessOutputMap[RaybotCommandType]
  PENDING: undefined
  IN_PROGRESS: undefined
}

export interface RaybotCommand<
  T extends RaybotCommandType = RaybotCommandType,
  S extends RaybotCommandStatus = RaybotCommandStatus,
> {
  id: string
  raybotId: string
  type: T
  status: RaybotCommandStatus
  input: InputMap[T]
  output: OutputMap[S]
  createdAt: Date
  completedAt?: Date
}
